Add a frontend for Quokka dataset

chongchonghe opened the pull request:
## PR Summary

This PR add a frontend for loading Quokka dataset. This frontend fully support uniform-grid or AMR datasets in cartesian coordinates and particles. QUOKKA is based on the AMReX framework, so we define a new `QuokkaDataset` class that inherits from `AMReXDataset`. 

This PR is fully tested on my computer. We have also added Answer Tests in `frontends/amrex/test_outputs.py` to validate new features with Quokka data that we provided. 

We prepared a Jupyter Notebook with examples and test dataset: https://github.com/Rongjun-ANU/README-of-yt-frontend-for-QUOKKA/blob/main/README.ipynb 

Authors: @Rongjun-ANU @chongchonghe 

[Quokka](https://github.com/quokka-astro/quokka) is a two-moment AMR radiation hydrodynamics (with self-gravity, particles, and chemistry) on CPUs/GPUs for astrophysics. 

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding , or if you would like help in
addressing a reviewer's comments.  And please ping us if you've been waiting
too long to hear back on your PR.-->

chongchonghe commented:
@BenWibking @Rongjun-ANU

chongchonghe commented:
I couldn't understand the remaining pre-commit.ci failsures. I would appreciate some explanation or help from the reviewers.

chongchonghe commented:
pre-commit.ci autofix

chrishavlin commented on doc/source/examining/loading_data.rst:
```diff
@@ -2133,6 +2133,175 @@ zero.
+Face-Centered Variables
+^^^^^^^^^^^^^^^^^^^^^^^
+
+QUOKKA datasets can also include face-centered variables. These are stored in
+the `fc_vars` directory which is inside the plotfile directory. When loading a
+Quokka dataset, YT automatically detects and loads any face-centered variables
+from the `fc_vars/x*****`, `fc_vars/y*****`, and `fc_vars/z*****`
+subdirectories. These are made available as attributes of the main dataset
+(`ds.ds_fc_x`, `ds.ds_fc_y`, `ds.ds_fc_z`). Each face-centered dataset is a full
+YT dataset object with its own fields and mesh structure, and includes a
+reference back to the parent dataset, `ds.ds_fc_x.parent_ds`. This allows for
+easy access to both cell-centered and face-centered data in the same analysis
+workflow.
```
These lines are the ones pre-commit has flagged: I believe the error is just that your in-line code snippets here need to use two back ticks rather than one. e.g., 

```
    `fc_vars`
```

needs to be 
```
    ``fc_vars``
```

chrishavlin replied:
I think the rest of the pre-commit errors should be handled by the bot, but you may need to manually fix these first?

chongchonghe replied:
Thanks a lot for the help. I have fixed all the pre-commit errors.

chongchonghe commented:
pre-commit.ci autofix

chrishavlin commented:
@chongchonghe I should be able to take a detailed look late next week. 

one quick question: do you have a sample dataset you can share and upload? I think it'd be particularly helpful to have one that has the face-centered data as well. The way you've set up this frontend to have a nested dataset if the face-centered data exists is intriguing. But it may introduce some confusing bugs -- the yt dataset-related classes rely on many class attributes and loading multiple datasets of the same class at once can lead to errors (e.g., like the one that PR #5159 fixed). So it'd be good if we can thoroughly test that case. 

Usually, we try to get sample datasets uploaded to the yt website so that new frontend tests can utilize the data (usually done via a pull request to https://github.com/yt-project/website like [this one](https://github.com/yt-project/website/pull/129) or [this one](https://github.com/yt-project/website/pull/128), but let me check back with you before you do so. The server that the full test suite runs on is having some issues, so I'm not sure we'll follow the usual protocol right now... at the very least it'd be useful if you could share some datasets so that I can download and run the new test(s) locally.

chongchonghe commented:
@chrishavlin 

This directory contains face-centered Quokka dataset (`sample/HydroWave`) and a script (`test.py`) to test that:
https://github.com/Rongjun-ANU/README-of-yt-frontend-for-QUOKKA . Running `test.py` using the YT from this PR is successful on my computer.

chongchonghe commented:
Unfortunately, I couldn't get pytest to work on my computer. I always get this error: 

```
E   ModuleNotFoundError: No module named 'imp'
```

It seems like the answer_testing in yt is using nose, which relies on imp, which is deprecated. However, I run the new tests I added in this PR on my computer as a python script and it worked, so I'm sure the unit tests will pass once we upload the dataset to the server.

chrishavlin commented:
thanks for the dataset! hopefully will have time to try things out later this week.

chrishavlin commented:
Hi @chongchonghe thanks for the ping, but unfortunately not yet. Had to bump it to this week's to-do list .

chrishavlin commented on pyproject.toml:
```diff
@@ -121,6 +121,7 @@ open-pmd = ["yt[HDF5]"]
 owls = ["yt[HDF5]"]
 owls-subfind = ["yt[HDF5]"]
 parthenon = ["yt[HDF5]"]
+quokka = ["PyYAML>=6.0.1"]
```
You can chain dependencies here to avoid defining dependencies in multiple places:

```
quokka = ["yt[amrex]"]
```

Note that if we decide to add PyYAML as a quokka-only dependency (see my main review comment), this line would be

```
quokka = ["yt[amrex]", "PyYAML>=6.0.1"]
```

and you'd need to change the `amrex` line back to:

```
amrex = []
```

chrishavlin reviewed (commented):
So I'm still working through the code, but wanted to bring up a high-level question. 

Normally, when frontends introduce dependencies, those dependencies should be added only to the frontend that requires them. But right now you're making PyYaml a dependency of all the amrex subclass frontends. It's a pretty small dependency, so I'm inclined to let it go in this instance, especially since it's already a dependency for a full yt install (via pyaml)... @neutrinoceros do you have an opinion on this? 

For reference, here's a diff that would make PyYaml a dependency only for Quokka (adjusting pyproject.toml, moving the imports to the functions where they're used):

<details>

<summary>Diff is here</summary>

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 03300c836..16a5f9b41 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -89,7 +89,7 @@ Fortran = ["f90nml>=1.1"]
 # We also normalize all target names to lower case for consistency.
 adaptahop = []
 ahf = []
-amrex = ["PyYAML>=6.0.1"]
+amrex = []
 amrvac = ["yt[Fortran]"]
 art = []
 arepo = ["yt[HDF5]"]
@@ -121,7 +121,7 @@ open-pmd = ["yt[HDF5]"]
 owls = ["yt[HDF5]"]
 owls-subfind = ["yt[HDF5]"]
 parthenon = ["yt[HDF5]"]
-quokka = ["PyYAML>=6.0.1"]
+quokka = ["yt[amrex]", "PyYAML>=6.0.1"]
 ramses = ["yt[Fortran]", "scipy"]
 rockstar = []
 sdf = ["requests>=2.20.0"]
diff --git a/yt/frontends/amrex/data_structures.py b/yt/frontends/amrex/data_structures.py
index 7a27f4efb..4e584de6b 100644
--- a/yt/frontends/amrex/data_structures.py
+++ b/yt/frontends/amrex/data_structures.py
@@ -6,7 +6,6 @@ from functools import cached_property
 from stat import ST_CTIME
 
 import numpy as np
-import yaml
 
 from yt.data_objects.index_subobjects.grid_patch import AMRGridPatch
 from yt.data_objects.static_output import Dataset
@@ -1373,6 +1372,7 @@ class QuokkaDataset(AMReXDataset):
     _field_info_class = QuokkaFieldInfo
     _subtype_keyword = ""
     _default_cparam_filename = "metadata.yaml"
+    _load_requirements = ["yaml"]
 
     def __init__(
         self,
@@ -1465,6 +1465,8 @@ class QuokkaDataset(AMReXDataset):
                 mylog.debug(f"No face-centered {direction} dataset found at {fc_dir}")
 
     def _parse_parameter_file(self):
+        import yaml
+
         # Call parent method to initialize core setup by yt
         super()._parse_parameter_file()
 
@@ -1686,6 +1688,8 @@ class QuokkaDataset(AMReXDataset):
         mylog.debug("Parsed header parameters: %s", self.parameters)
 
     def _parse_metadata_file(self):
+        import yaml
+
         # Construct the full path to the metadata file
         metadata_filename = os.path.join(self.output_dir, self.cparam_filename)
         try:
```

</details>

chrishavlin commented:
> Now I'm in favor of committing your patch :) 

Hah, ok, then let's go with that. @chongchonghe , could you update the code so that `pyyaml` is only needed for quokka (check the diff in my previous comment)?

chongchonghe commented:
@chrishavlin I've implemented the changes as you suggested. Thanks a lot!

I have also been trying to optimize how we deal with face-centering variables. I came to the conclusion that the current approach is the optimal approach. QUOKKA does not currently support writing FC variables as 'raw' data as WarpX does. So, I did a hack by loading the face-centered quantities as if they are cell-centered. Raw and "nodal" fields will be supported in the future. 

Sorry for my delayed response. I've been overwhelmed with a proposal recently. Please let me know if any revisions are required, or if you need any additional information from me before we can proceed with merging this PR.

neutrinoceros commented:
I am not currently available to review a large PR as this one, I apologize for the inconvenience.

chrishavlin commented on yt/frontends/amrex/tests/test_outputs.py:
```diff
@@ -445,6 +446,96 @@ def test_maestro_parameters():
     assert type(ds.parameters["s0_interp_type"]) is int  # noqa: E721
 
 
+quokka = "quokka_RadiatingParticles_plt026"
```
Given how you have https://github.com/Rongjun-ANU/README-of-yt-frontend-for-QUOKKA/tree/main set up, it will be easier to just use paths here: 

```suggestion
quokka = "quokka/RadiatingParticles/plt026"
```

Then the `sample` folder from https://github.com/Rongjun-ANU/README-of-yt-frontend-for-QUOKKA/tree/main can be copied to a `quokka` directory within yt's test data directory (rather than renaming, e.g., `sample/RadiatingParticles/plt026` to `sample_RadiatingParticles_plt026`).

chrishavlin commented on yt/frontends/amrex/tests/test_outputs.py:
```diff
@@ -445,6 +446,96 @@ def test_maestro_parameters():
+    assert particle_info["units"]["end_time"] == "T^1"
+
+
+quokka_face_centered = "quokka_HydroWave_plt00004"
```
```suggestion
quokka_face_centered = "quokka/HydroWave/plt00004"
```

chrishavlin reviewed (commented):
Thanks for your patience @chongchonghe ! I have some minor suggestions here to clean up some of the new code. I'll also see if anyone else is available for a quick review since we like to have at least 2 reviews for new frontends: but I recognize this has been around for quite a while and it'd be nice to get it merged ASAP. 

And I did download the test data and confirmed that the new tests run successfully. I'll make a companion PR to the yt website repo with the data in a convenient form so it ends up where it needs to eventually, but note that the tests won't actually run for now while yt's continuous integration server is down.

chongchonghe commented:
> hey @chongchonghe , i'm drumming up support for a yt feature release (version 4.5.0) in January, it'd be great to get this merged so it's included in the release! Do you mind fixing the merge conflict and checking out my last batch of comments? they're pretty minor and should be quick. 

Sure. I'll do it today.

chongchonghe commented:
> hey @chongchonghe , i'm drumming up support for a yt feature release (version 4.5.0) in January, it'd be great to get this merged so it's included in the release! Do you mind fixing the merge conflict and checking out my last batch of comments? they're pretty minor and should be quick.

hi @chrishavlin . Sorry for the late response. I have addressed the comments and resolved the merge conflicts. Can you review and approve this PR?

BenWibking commented on tests/.DS_Store:
This should be removed.

matthewturk reviewed (commented):
This is outstanding work.  Thank you -- and I'm sorry that I didn't review this much sooner.  I have some comments, and while I'd like to see them applied, I don't believe any rise to the level of blockers.

I do think we'll need to get the `uv.lock` situation fixed, but I'm not 100% sure if that's trivial or not.

chongchonghe commented:
pre-commit.ci autofix

chongchonghe commented on yt/frontends/amrex/fields.py:
```diff
@@ -504,6 +504,258 @@ def setup_fluid_fields(self):
+        # Scalar fields are not always present
+        ("scalar_0", ("", ["scalar_0"], "Scalar 0")),
+        ("scalar_1", ("", ["scalar_1"], "Scalar 1")),
+        ("scalar_2", ("", ["scalar_2"], "Scalar 2")),
```
I have replaced it with `self.setup_scalar_fields()`

neutrinoceros commented:
given that at least one file (`.DS_Store`) was accidentally added then removed, and we don't want it to live forever in yt's commit history, please squash your branch history with `git rebase`.

chongchonghe commented on yt/frontends/amrex/data_structures.py:
```diff
@@ -1327,10 +1330,404 @@ def _set_code_unit_attributes(self):
+
+                # Check and parse Fields.yaml
+                if os.path.exists(fields_yaml_file):
+                    try:
```
ok, i use a `ytcfg.get("yt", "quokka_debug")` . i think that is cleaner than a global variable DEBUG_QUOKKA

neutrinoceros commented:
> I thought that if we delete a branch after Squash and Merge to main branch, it will delete the unused files from git history. Is this not true?

that is correct, however it's very easy for maintainers to forget using Squash-rebase when the time comes so I'd rather we do a rebase now.

chongchonghe committed:
Squashed commit of the following:

commit 7d07c2f7a349cb792e397bbe24bc9d3a547b3683
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Wed Mar 25 13:27:08 2026 +1100

    uv lock

commit bfea253035683ce4c3ecc4878ad6e7cce4606031
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Mon Mar 23 05:07:45 2026 +1100

    _load_requirements

commit 0fb09209d9660bdbca544daed3a5acfc02cdb37c
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Mon Mar 23 05:05:39 2026 +1100

    comment

commit e455e1f331d84f2586f64350385eddb3ba4648b7
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Mon Mar 23 04:50:50 2026 +1100

    refactor Quokka particle parsing and replace DEBUG_QUOKKA with ytcfg

    Replace the module-level DEBUG_QUOKKA flag with ytcfg.get("yt",
    "quokka_debug") so the debug mode can be toggled at runtime.

    Also fix a bug where a particle type was added to
    detected_particle_types but its entry was skipped in particle_info
    when Fields.yaml was absent (inverted early-continue). Additional
    cleanup: remove dead yaml_field_names=None init, raw_units alias,
    and unreachable real_comp* loop; extract _QUOKKA_UNIT_DIMS as a
    module-level constant; simplify unit builder with an if-guard in
    the generator and a dict comprehension.

commit 3e4dde0bb36dfbd314429d8a3a26167958cc1884
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Mon Mar 23 04:37:07 2026 +1100

    deindent

commit 075a9dcb3f4d271e59f445d26d4f7579bb2c5a11
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Mon Mar 23 04:29:03 2026 +1100

    config

commit 5fb96550af168c0fc0e02738b81dfb3bab278310
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Mon Mar 23 02:00:12 2026 +1100

    fix bugs introcued by API change in main branch

commit 612c4ff6df9bdbfedbfd283ec6179256159ea8a1
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Mon Mar 23 00:26:42 2026 +1100

    list

commit cd10da40bfd56eb2d3cc3527f1845d5de6570560
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Sun Mar 22 13:51:54 2026 +1100

    Address PR review comments for Quokka frontend

    - Remove tests/.DS_Store
    - Replace yt.load import with direct AMReXDataset instantiation to avoid circular imports
    - Use on_demand_imports (_yaml) for yaml instead of inline imports; add safe_load and YAMLError to yaml_imports
    - Add DEBUG_QUOKKA flag to allow re-raising exceptions during FC/metadata loading
    - Invert face-centered if-logic to reduce nesting (early continue pattern)
    - Use sorted(particle_dirs) for deterministic ordering
    - Support arbitrary number of scalar fields dynamically (like radiation groups)
    - Count scalar_field_count in _parse_parameter_file
    - Restructure _parse_metadata_file try/except to wrap only yaml.safe_load

commit d3996ad240e00c85660078612b5a3887ad4878d0
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Sun Mar 22 13:00:42 2026 +1100

    pre-commit

commit 616849ea7cd398c5c5b912c5ff624c78f32077f3
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Sun Mar 22 12:48:59 2026 +1100

    uv lock

commit f167e3bbff0f6a5291bb304955dd068da0f9ca54
Merge: 2bbc267ca e512d14d6
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Sun Mar 22 12:45:52 2026 +1100

    Merge branch 'main' into Rongjun-ANUquokka-frontend

commit 2bbc267caaa0c716d820b2e9d6bbe4c7b457ee4c
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Thu Jan 1 11:44:49 2026 +0800

    pre-comit

commit 09a91e8f1fac94356091988be6e2639e9ee4b39f
Merge: f41a18272 a1ddf6567
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Thu Jan 1 11:34:03 2026 +0800

    Merge remote-tracking branch 'upstream/main' into Rongjun-ANUquokka-frontend

commit f41a18272effbc4a5d9b3a5426f9600a69aa00ff
Author: ChongChong He <chongchonghe99@gmail.com>
Date:   Thu Jan 1 11:23:45 2026 +0800

    address comments

commit 581ad28a563e4789ac53b6138232c2307351deb0
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Wed Jun 11 00:26:01 2025 +1000

    make PyYAML a dependency of Quokka only

commit d6799ce6fa02dcf45998a43bb3199964436ca2a2
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Sun May 18 12:18:16 2025 +1000

    whitespace

commit dfc69a821a6f2dcfa65ec7fcfecee2d4d28ab55f
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Sun May 18 12:14:50 2025 +1000

    rename dataset name

commit 75bfe3f7f0375aa8f4aa0a7e66946e0ef1f2b4e4
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Sun May 18 11:21:47 2025 +1000

    add tests of face-centered data

commit 1542a026a3d2d30256cf10a4e081e704069c5535
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Fri May 16 12:47:24 2025 +1000

    pre-commit run all

commit 145d0eac4be89e6ad64d3b722113936f1923bcdd
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Fri May 16 12:42:07 2025 +1000

    fix style

commit 9997397b58f9ec3be83b37e67e8aadd32ae39c2b
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Fri May 16 12:39:07 2025 +1000

    fix style issue

commit bf56f910a6459fce0edcce79cb54297711b97dc7
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Thu May 15 17:16:24 2025 +1000

    remove trailing white space

commit 8935ab795e9314a2fc62b8a20a0d0a227a13c571
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Thu May 15 17:12:49 2025 +1000

    fix lambda error

commit fa2289a9e9a81fe516236a137f0d5b8d74ebe7e8
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Thu May 15 17:06:39 2025 +1000

    fix pre-commit.ci errors

commit 225041047b2b07fb442716e6b26fad0185fc2927
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Thu May 15 16:35:58 2025 +1000

    format

commit 60db2cfa305d7e6e14897e6f3ea82de3c108667a
Merge: 5ad1e7828 dbf530510
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Thu May 15 16:17:40 2025 +1000

    Merge branch 'main' into Rongjun-ANUquokka-frontend

commit 5ad1e7828fb5eaa54eeb9582e43ce2549ec4cb0d
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Thu May 15 16:05:27 2025 +1000

    Add face-centered dataset support (#20)

    Add support for reading face-centered dataset. See changes in
    loading_data.rst for details.

    <!--Please provide at least 1-2 sentences describing the pull request in
    detail.  Why is this change required?  What problem does it solve?-->

    <!--If it fixes an open issue, please link to the issue here.-->

    <!-- Note that some of these check boxes may not apply to all pull
    requests -->
    - [x] New features are documented, with docstrings and narrative docs
    - [ ] Adds a test for any bugs fixed. Adds tests for new features.

    <!--We understand that PRs can sometimes be overwhelming, especially as
    the
    reviews start coming in. Please let us know if the reviews are unclear
    or the
    recommended next step seems overly demanding , or if you would like help
    in
    addressing a reviewer's comments. And please ping us if you've been
    waiting
    too long to hear back on your PR.-->

commit 932b7c46d8c545caad9ec30ea45a2e9ee2b0454e
Author: Rongjun Huang <u6569836@anu.edu.au>
Date:   Sun Apr 27 16:08:56 2025 +1000

    test_quokka (#19)

    update the `yt/frontends/amrex/tests/test_outputs.py` to run pytest for
    quokka dataset

    ---------

    Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>
    Co-authored-by: ChongChong He <chongchong.he@anu.edu.au>

commit 075769d3774ff47adb7ba571a18be0ace3d790d0
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Mon Mar 24 15:04:15 2025 +1100

    Chong/update-loading_data-rst (#17)

    update YT docs for Quokka frontend

commit 7755781916d9be511b4c50fe4e26842b68d0fea2
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Mon Mar 24 11:34:55 2025 +1100

    add PyYAML to amrex (#16)

    <!--Please provide at least 1-2 sentences describing the pull request in
    detail.  Why is this change required?  What problem does it solve?-->

    <!--If it fixes an open issue, please link to the issue here.-->

    <!-- Note that some of these check boxes may not apply to all pull
    requests -->
    - [ ] New features are documented, with docstrings and narrative docs
    - [ ] Adds a test for any bugs fixed. Adds tests for new features.

    <!--We understand that PRs can sometimes be overwhelming, especially as
    the
    reviews start coming in. Please let us know if the reviews are unclear
    or the
    recommended next step seems overly demanding , or if you would like help
    in
    addressing a reviewer's comments. And please ping us if you've been
    waiting
    too long to hear back on your PR.-->

commit 9c1103d66a0a82ae36310691affd69f80b4bcc41
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Fri Mar 21 22:44:46 2025 +1100

    supported structured metadata.yaml (#15)

    Recently I have updated the format of metadata.yaml from Quokka
    simulations, and also include `quokka_version` variable in
    metadata.yaml. See https://github.com/quokka-astro/quokka/pull/935 .
    This PR updates the Quokka frontend to support this new, nested yaml
    metadata.

    <!-- Note that some of these check boxes may not apply to all pull
    requests -->
    - [ ] New features are documented, with docstrings and narrative docs
    - [ ] Adds a test for any bugs fixed. Adds tests for new features.

    <!--We understand that PRs can sometimes be overwhelming, especially as
    the
    reviews start coming in. Please let us know if the reviews are unclear
    or the
    recommended next step seems overly demanding , or if you would like help
    in
    addressing a reviewer's comments. And please ping us if you've been
    waiting
    too long to hear back on your PR.-->

commit 8aad9dcf81270841440b3fc838a8d2b1923d8676
Author: Rongjun Huang <u6569836@anu.edu.au>
Date:   Thu Mar 20 22:39:04 2025 +1100

    Remove rebundancy (#14)

    Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>

commit 5988844bc8654621066e8a92de473e0b2c0adc9e
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Thu Mar 20 19:08:04 2025 +1100

    add quokka = ["PyYAML>=6.0.1"] (#13)

    add quokka = ["PyYAML>=6.0.1"] (#13)

commit 0fefa614e932579140283f346f8ea59bea3e2762
Author: Rongjun Huang <u6569836@anu.edu.au>
Date:   Thu Mar 20 15:10:59 2025 +1100

    rename Fields.json to Fields.yaml

    rename Fields.json to Fields.yaml

    ---------

    Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>

commit c7ad3e60f65e62b3ba689a7156f43f9b4dc5db12
Author: Rongjun Huang <u6569836@anu.edu.au>
Date:   Wed Jan 15 01:34:57 2025 +1100

    Rongjun read quokka metadata header (#10)

    update for Bfields

    ---------

    Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>

commit 7910f07f507a4e836dcc7c612a97409edf1e6267
Merge: 6b455426a a887cf012
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Tue Jan 14 16:48:58 2025 +1100

    Merge branch 'main' into Rongjun-ANUquokka-frontend

commit 6b455426a959fa967a70fb254759f2f9c05f38c1
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Tue Jan 14 16:40:12 2025 +1100

    more cleanup

commit cdb199775e5e9f70fb3fe944b850b8b438ab53eb
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Tue Jan 14 16:30:19 2025 +1100

    clean up

commit ae0dd250fcd113016fbad30d17d42aabe6cb65a3
Author: Rongjun Huang <u6569836@anu.edu.au>
Date:   Tue Jan 7 11:23:08 2025 +1100

    further support for particles and their units (#8)

    add support for `Fields.json` in particle folders. also readout the unit of particles.

    ---------

    Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>

commit bd7dc58568550770387c4e0fc84f0aac52dd3d9d
Merge: f5acf45ef 8c467da65
Author: ChongChong He <chongchong.he@anu.edu.au>
Date:   Tue Dec 17 17:34:54 2024 +1100

    Merge branch 'main' into Rongjun-ANUquokka-frontend

commit f5acf45ef1f936bbab5149678136457d29197ee8
Author: Rongjun Huang <u6569836@anu.edu.au>
Date:   Tue Dec 10 23:51:24 2024 +1100

    add support of multiple particle types (#6)

    slight update on `class QuokkaHierarchy(BoxlibHierarchy)` to fix the
    bug, and also add support on multi particle types.

    ---------

    Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>

commit cd0dab6433fce453a7d10b08d619b194c88772cf
Author: Rongjun Huang <u6569836@anu.edu.au>
Date:   Tue Dec 10 16:36:20 2024 +1100

    Rongjun update support for temperature, scalars, rad, velocities, BFields fields (#5)

    This pull request introduces support for the Quokka simulation framework
    by adding new dataset and hierarchy classes, and updating field
    information. The most important changes include the addition of
    `QuokkaDataset` and `QuokkaHierarchy` classes, the parsing of metadata
    and parameter files, and the setup of fluid and radiation fields.

    * Added `QuokkaDataset` and `QuokkaHierarchy` classes to handle
    Quokka-specific data structures and hierarchy.
    (`yt/frontends/amrex/data_structures.py`,
    [yt/frontends/amrex/data_structures.pyR1332-R1571](diffhunk://#diff-7aa59ffaeb7262d9580ae294f19155633ce68e2dda607cc4e2ddbf1bc20b66b3R1332-R1571))
    * Implemented `_parse_parameter_file` and `_parse_metadata_file` methods
    in `QuokkaDataset` to read and parse metadata from `metadata.yaml` and
    other parameter files. (`yt/frontends/amrex/data_structures.py`,
    [yt/frontends/amrex/data_structures.pyR1332-R1571](diffhunk://#diff-7aa59ffaeb7262d9580ae294f19155633ce68e2dda607cc4e2ddbf1bc20b66b3R1332-R1571))

    * Added `QuokkaFieldInfo` class to define known fields and set up fluid
    and radiation fields dynamically. (`yt/frontends/amrex/fields.py`,
    [yt/frontends/amrex/fields.pyR506-R597](diffhunk://#diff-565dd12ba00274995d4c360bd7cb3f7d43068a0f2c7f93fe2d33978c1e90c98fR506-R597))

    * Included `QuokkaDataset`, `QuokkaHierarchy`, and `QuokkaFieldInfo` in
    the API imports to make them accessible. (`yt/frontends/amrex/api.py`,
    [[1]](diffhunk://#diff-92252639991f47b99474012776b030614e644789c56c7a9c39661ee5a301f87bR14-R15)
    [[2]](diffhunk://#diff-92252639991f47b99474012776b030614e644789c56c7a9c39661ee5a301f87bR24)

    * Added `yaml` import for parsing metadata files.
    (`yt/frontends/amrex/data_structures.py`,
    [yt/frontends/amrex/data_structures.pyR1](diffhunk://#diff-7aa59ffaeb7262d9580ae294f19155633ce68e2dda607cc4e2ddbf1bc20b66b3R1))

    ---------

    Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>

commit eacc9b75d00e5abd49d11c868acf8299ed9930e3
Author: Rongjun Huang <u6569836@anu.edu.au>
Date:   Wed Nov 20 18:24:48 2024 +1100

    Add Quokka dataset and hierarchy support with metadata and header parsing (#3)

    Updated the following classes to read header file and `metadata.yaml` in
    Quokka simulations.

    ```
    class QuokkaHierarchy(BoxlibHierarchy)
    class QuokkaDataset(AMReXDataset)
    ```

    ---------

    Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>

chongchonghe committed:
Add a frontend for Quokka datasets

Co-authored-by: Rongjun-ANU <Rongjun.Huang@alumni.anu.edu.au>
Co-authored-by: ChongChong He <chongchong.he@anu.edu.au>

chongchonghe commented:
OK, done. All squashed into a single commit.

chongchonghe committed:
uv lock

matthewturk reviewed (dismissed):
I approved!

chongchonghe committed:
merge

chongchonghe committed:
merge in uv.lock

chongchonghe commented:
> I think that both the failures on 3.10 and the conflict on uv.lock are symptoms of a single problem, which you _should_ be able to resolve by rebasing on the current state of main. I would advise dropping the commit on uv.lock in the process, as I think it's slightly malformed (`--exclude-newer` wasn't used). I can fix this for you (including the entire rebase), though I feel uncomfortable force pushing to your branch without so much as a warning, so I'll give you a couple days at least to respond and/or do the rebase yourself.

Please feel free to do any kind of uv lock tweaks and push to this PR. I don't see any problem for a maintainer to do this kind of work to make the PR comply with the code requirement. I would also appreciate a contributing guidelines or a PRD template that explains the uv lock trick.

chongchonghe committed:
fix bug, should pass frontend tests

neutrinoceros commented on yt/config.py:
```diff
@@ -43,6 +43,7 @@
     "supp_data_dir": "/does/not/exist",
     "default_colormap": "cmyt.arbre",
     "ray_tracing_engine": "yt",
+    "quokka_debug": False,
```
There's no precedent for this

```suggestion
```

chongchonghe replied:
OK, i replaced it with `mylog.isEnabledFor(logging.DEBUG)`

chrishavlin replied:
resolved! yt.config is no longer being used here.

neutrinoceros reviewed (dismissed):
I finally pulled off this difficult rebase. For whatever reason, my fork seems to be stuck behind, but I don't think I need to fix this right now.
I ran a 2-min review and found patterns I'd really like to avoid having on stable.

chongchonghe committed:
Merge branches 'Rongjun-ANUquokka-frontend' and 'Rongjun-ANUquokka-frontend' of github.com:chongchonghe/yt into Rongjun-ANUquokka-frontend

chongchonghe committed:
Merge remote-tracking branch 'upstream/main' into Rongjun-ANUquokka-frontend

chongchonghe committed:
LOGGING

chongchonghe committed:
remove print

chongchonghe committed:
remove print

chongchonghe committed:
merge

chongchonghe commented on yt/frontends/amrex/data_structures.py:
```diff
@@ -1327,10 +1333,385 @@ def _set_code_unit_attributes(self):
+        mylog.debug("Header loaded successfully")
+        mylog.debug("Parsed header parameters: %s", self.parameters)
+
+    def _parse_metadata_file(self):
```
OK, this makes sense. I've remove try/except and use proper error throw

chongchonghe committed:
type: BField

chongchonghe committed:
fix particle num

chongchonghe commented:
@neutrinoceros @matthewturk I addressed the comments, cleaned up some code, and tested it locally. Everything looks fine. Can you review and approve again?

matthewturk commented:
I think this is good to go!

chongchonghe committed:
fix bug: integer

chongchonghe committed:
make particle components read more robust

chongchonghe commented:
@matthewturk Thank! I made a minor change to adopt to a recent update on particle component names in quokka. Can you approve again? Also, is there anything I need to do before you merge it?

BenWibking commented:
We are preparing to release Quokka 26.04 in the next few days, and would like to mention the addition of a full-featured frontend in yt to our users. If it's possible to re-review and merge this in that timeframe, please let us know.

neutrinoceros commented:
In order to harmonize expectations, I'm letting you know I don't have time to review this properly in the forseeable future, so I'm removing myself from review requests here. I trust other maintainers to make sure my existing comments were actually addressed before merging.

chongchonghe commented:
@matthewturk @chrishavlin Can you approve and merge this PR if there is no further request of change?

chrishavlin commented:
>I trust other maintainers to make sure my existing comments were actually addressed before merging. 

@neutrinoceros from my quick look at the end of last week, the comments were addressed. I'll double check now...

chrishavlin reviewed (approved):
Thanks for all your hard work and patience on this! All the comments are now addressed.

chrishavlin commented:
Ok! let's do this... thanks again @chongchonghe @BenWibking @matthewturk @neutrinoceros !
